server/app/auth/oidc/OidcProvider.java

```diff
@@ -0,0 +1,185 @@
+
+  private static final Logger logger = LoggerFactory.getLogger(OidcProvider.class);
+  protected final Config configuration;
+  protected final ProfileFactory profileFactory;
+  protected final Provider<UserRepository> applicantRepositoryProvider;
+  protected final String baseUrl;
+
+  public OidcProvider(
+      Config configuration,
+      ProfileFactory profileFactory,
+      Provider<UserRepository> applicantRepositoryProvider) {
+    this.configuration = checkNotNull(configuration);
+    this.profileFactory = checkNotNull(profileFactory);
+    this.applicantRepositoryProvider = checkNotNull(applicantRepositoryProvider);
+
+    this.baseUrl =
+        getBaseConfigurationValue("base_url")
+            .orElseThrow(() -> new RuntimeException("base_url must be set"));
+  }
+
+  /*
+   * The prefix used in the application.conf for retriving oidc options.
+   */
+  protected abstract String attributePrefix();
+
+  /*
+   * The OIDC Provider Name (optional).
+   */
+  protected abstract Optional<String> getProviderName();
+
+  /*
+   * Provide the profile adaptor that should be used.
+   */
+  public abstract ProfileCreator getProfileAdapter(OidcConfiguration config, OidcClient client);
+
+  /*
+   * The OIDC Client ID.
+   */
+  protected abstract String getClientID();
+
+  /*
+   * The OIDC Client Secret.
+   */
+  protected abstract String getClientSecret();
+
+  /*
+   * The OIDC Discovery URI.
+   */
+  protected abstract String getDiscoveryURI();
+
+  /*
+   * The OIDC Response Mode.
+   */
+  protected abstract String getResponseMode();
+
+  /*
+   * The OIDC Response Type.
+   */
+  protected abstract String getResponseType();
+
+  /*
+   * Default scopes for this OIDC implementation.
+   */
+  protected abstract ImmutableList<String> getDefaultScopes();
+
+  /*
+   * Any application.conf-defined extra scopes for this OIDC implementation.
+   */
+  protected abstract ImmutableList<String> getExtraScopes();
+
+  /*
+   * Helper function for retriving values from the application.conf,
+   * prepended with "<attributePrefix>."
+   */
+  protected final Optional<String> getConfigurationValue(String suffix) {
+    String name = attributePrefix() + "." + suffix;
+    return getBaseConfigurationValue(name);
+  }
+
+  /*
+   * Helper function for retriving values from the application.conf,
+   * prepended with "<attributePrefix>."
+   */
+  protected final Optional<String> getBaseConfigurationValue(String name) {
+    if (configuration.hasPath(name)) {
+      return Optional.ofNullable(configuration.getString(name));
+    }
+    return Optional.empty();
+  }
+
+  protected String getCallbackURL() {
+    return baseUrl + "/callback";
+  }
+
+  private final String getScope() {
+    // Scopes are the other things that we want from the OIDC endpoint
+    // (needs to also be configured on provider side).
+    ImmutableList<String> allClaims =
+        ImmutableList.<String>builder().addAll(getDefaultScopes()).addAll(getExtraScopes()).build();
```

Comment:
Blocking: What if getExtraScopes / getDefaultScopes have duplicate values? Presumably we'd specify the scope multiple times in the result string. Does OIDC gracefully handle that?

Assuming order doesn't matter, I'd suggest adding these to a set in order to prevent this type of issue.

Reply:
done